json-schema-processor/src/main/java/io/micronaut/jsonschema/visitor/JsonSchemaVisitor.java

```diff
@@ -0,0 +1,220 @@
+            if (context == null) {
+                context = JsonSchemaContext.createDefault();
+                visitorContext.put(JSON_SCHEMA_CONFIGURATION_PROPERTY, context);
+            }
+            createTopLevelSchema(element, visitorContext, context);
+        }
+    }
+
+    public static Schema createTopLevelSchema(TypedElement element, VisitorContext visitorContext, JsonSchemaContext context) {
+        Schema schema = context.createdSchemasByType().get(element.getGenericType().getName());
+        if (schema != null) {
+            return schema;
+        }
+        schema = new Schema();
+
+        AnnotationValue<JsonSchema> schemaAnn = element.getGenericType().getDeclaredAnnotation(JsonSchema.class);
+        if (schemaAnn != null) {
+            schema.setTitle(schemaAnn.stringValue("title")
+                .orElse(element.getGenericType().getSimpleName().replace('$', '.')));
+            schemaAnn.stringValue("description").ifPresent(schema::setDescription);
+
+
+            Optional<String> uriOptional = schemaAnn.stringValue("uri");
+            String uri;
+            if (uriOptional.isPresent()) {
+                uri = uriOptional.get();
+                if (!uri.contains("://")) {
+                    uri = uri + SUFFIX;
+                }
+            } else {
+                uri = SLASH + NameUtils.camelCaseToKebabCase(schema.getTitle()) + SUFFIX;
+            }
+            if (!uri.contains("://")) {
+                if (context.baseUrl() != null) {
+                    uri = context.baseUrl() + uri;
+                } else {
+                    visitorContext.warn("The JSON schema for type " + element.getName()
+                        + " does not have a resolvable URI", element);
+                }
+            }
+            schema.set$id(uri);
+            schema.set$schema(context.draft().getDraftUrl());
+        }
+
+        setSchemaType(element, visitorContext, context, schema);
+
+        for (SchemaInfoAggregator aggregator: SCHEMA_INFO_AGGREGATORS) {
+            schema = aggregator.addInfo(element, schema, visitorContext, context);
+        }
+
+        if (schemaAnn != null) {
+            context.createdSchemasByType().put(element.getGenericType().getName(), schema);
+            writeSchema(schema, element.getGenericType(), visitorContext, context);
+        }
+        return schema;
+    }
+
+    public static Schema createSchema(TypedElement element, VisitorContext visitorContext, JsonSchemaContext context) {
+        Schema topLevelSchema = createTopLevelSchema(element, visitorContext, context);
+        if (topLevelSchema.get$id() != null) {
+            return Schema.reference(topLevelSchema.get$id());
+        }
+        return topLevelSchema;
+    }
+
+    private static void setSchemaType(TypedElement element, VisitorContext visitorContext, JsonSchemaContext context, Schema schema) {
```

Comment:
we probably need a way for the developer to manually specify the type

Reply:
Yes, but I don't think type would be enough if users would need to customize the schema. We probably need ability to customize other things, too. Other projects don't seem to offer customizing this field directly: https://github.com/FasterXML/jackson-module-jsonSchema, https://github.com/mbknor/mbknor-jackson-jsonSchema. Although the second repository supports `@JsonSchemaOptions` annotations which allows setting any property of the generated schema.